src/Nethermind/Nethermind.JsonRpc/Modules/RpcModuleProvider.cs

```diff
@@ -97,9 +97,9 @@ public ModuleResolution Check(string methodName, JsonRpcContext context)
 
             if ((result.Availability & context.RpcEndpoint) == RpcEndpoint.None) return ModuleResolution.EndpointDisabled;
 
-            if (context.Url != null) return context.Url.EnabledModules.Contains(result.ModuleType, StringComparer.InvariantCultureIgnoreCase) ? ModuleResolution.Enabled : ModuleResolution.Disabled;
+            if (context.Url != null) return context.Url.EnabledModules.Contains(result.ModuleType, StringComparer.InvariantCultureIgnoreCase) ? ModuleResolution.Enabled : ModuleResolution.EndpointDisabled;
```

Comment:
I'm sorry I wasn't specific enough. I meant more smth like this:
```
            if (context.Url != null && !context.Url.EnabledModules.Contains(result.ModuleType, StringComparer.InvariantCultureIgnoreCase))
            {              
                return ModuleResolution.EndpointDisabled;
            }
```
The last line returns if a module is enabled/disabled at all. The preceding lines indicate if the module is enabled/disabled at the context level.

Reply:
Shouldn't it determine enabled/disabled solely off of the context enabled modules if context.Url is not null? If I'm reading that logic right, I think it would end up hitting the last line to determine if it is enabled again using _enabledModules which comes from JsonRpcConfig.EnabledModules which is only there as a fallback for when Url is null (not http/websocket).

Reply:
We have this logic in JsonRpcService
```
            return result switch
            {
                ModuleResolution.Unknown => ((int?) ErrorCodes.MethodNotFound, $"Method {methodName} is not supported"),
                ModuleResolution.Disabled => (ErrorCodes.InvalidRequest, $"{methodName} found but the containing module is disabled, consider adding module to JsonRpcConfig.EnabledModule"),
                ModuleResolution.EndpointDisabled => (ErrorCodes.InvalidRequest, $"{methodName} found but disabled for {context.RpcEndpoint}"),
                _ => (null, null)
            };
```

Which indicates that the Disabled switch is linked with JsonRpcConfig.EnabledModules .
I can imagine that we have a module enabled in this config (EnabledModules) but it's not enabled for the URL - then we want to return EndpointDisabled .
**The question is how the system should behave if the module is NOT listed in EnabledModules but it's listed in the URL config. What is the priority?**
@MarekM25 , @LukaszRozmej  ?

Reply:
If it helps, the implementation was made with the requirement of keeping this backwards compatible, so JsonRpcConfig.Host/Port/WebSocketPort/EnabledModules  is used to create the first/default url in the url collection. This way the config will still work as it did before these changes if a user does not want to use the AdditonalRpcUrls functionality.

Reply:
Confirmed with @MarekM25 - both configurations (the old and the new one) are independent. So, it's good as it is - we can ignore my last suggestion :)

Reply:
OK cool. So the latest commit (changing it to return ModuleResolution.EndpointDisabled instead of Disabled) is correct?

Reply:
I'm just thinking that following the same logic, EndpointDisabled would imply that the specific endpoint is disabled (net.peerCount, but not net.listening), while in fact it is the entire module (net) which is what Disabled represents. If that is the case, the error message in JsonRpcService might need to be reworded.